### PR TITLE
TiledMLP support for FSDP2

### PR DESCRIPTION
**Pull request by winglian**
# Description
Use manual gradient accumulation with parameter hooks to work with FSDP2. I didn't test with FSDP1, but I expect it should work with that also. 
@sfc-gh-sbekman

WandB: https://wandb.ai/axolotl-ai/tiledmlp-fsdp2?nw=nwuserwingaxolotl

Screenshot 2025-07-19 at 1 30 11 PM

## Summary by CodeRabbit

* **New Features**
  * Introduced a new tiled multi-layer perceptron (TiledMLP) supporting distributed and single GPU training, with advanced gradient accumulation for mixed precision scenarios.
  * Added a new patch application step between plugin initialization and model loading to enhance model setup.
  * Added example configurations and documentation for Arctic Long Sequence Training (ALST) with Llama 3 8B models using DeepSpeed and FSDP.

* **Improvements**
  * Enhanced garbage collection during training to trigger at the very start and at step zero.
  * Updated tiled MLP patch to default to using the original MLP implementation and dynamically select DeepSpeed support when applicable.
  * Refined conflict validation to allow tiled MLP usage when explicitly configured to use the original MLP.
  * Changed default configuration to enable original MLP usage for tiled MLP by default.

* **Removals**
  * Removed the validation check that enforced DeepSpeed usage with tiled MLP in multi-GPU setups.

### CODE REVIEW

**Comment by winglian** [comment]
Screenshot 2025-07-19 at 1 44 13 PM
`HYBRID_SHARD` is accomplished using the 2d-mesh now

File: src/axolotl/monkeypatch/tiled_mlp/fsdp.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        x_requires_grad = x.requires_grad
+        x = x.detach()
+        x.requires_grad_(x_requires_grad)
+
+        incoming_grad = grads[0]
+        x_grad = torch.zeros_like(x)
+        x_shards = list(torch.chunk(x, chunks=shards, dim=1))
+
+        # Create a gradient accumulator for parameters
+        grad_accumulator = GradientAccumulator(compute_params, shards)
```
Comment (stas00) [inline src/axolotl/monkeypatch/tiled_mlp/fsdp.py]: Is there a way to detect this is running under FSDP?

You can detect it's running under deepspeed by checking `hasattr(param, "ds_param")` as deepspeed installs custom attributes into params.

I am thinking these two can then be merged into one class.

Also this would work for DDP as well, right?

Comment (winglian) [reply src/axolotl/monkeypatch/tiled_mlp/fsdp.py]: I couldn't consistently find the `ds_param` attribute, but `param_idx_in_group` seems to work across the various ZeRO-* implementations so it's still a single unified patch.

File: examples/alst/README.md
Comment (winglian) [inline examples/alst/README.md:1]: @stas00 I'm sure this could be better, so happy to accept any changes you have on this README.

File: README.md [hunk 2]
```diff
@@ -25,6 +25,7 @@
 
 ## 🎉 Latest Updates
 
+- 2025/07: TiledMLP support for single-GPU to multi-GPU training with DDP, DeepSpeed and FSDP support has been added to support Arctic Long Sequence Training. (ALST). See [examples](https://github.com/axolotl-ai-cloud/axolotl/tree/main/examples/alst) for using ALST with Axolotl!
```
Comment (salmanmohammadi) [inline README.md:28]:
```suggestion
- 2025/07: [Arctic Long Sequence Training](https://arxiv.org/abs/2506.13996) (ALST) has been integrated into Axolotl! Compose TiledMLP seamlessly with single-GPU training, and distributed training techniques such as DDP, DeepSpeed, and FSDP. Check out the [examples](https://github.com/axolotl-ai-cloud/axolotl/tree/main/examples/alst) to learn how to use ALST with Axolotl!
```

File: examples/alst/llama3-8b-fsdp2-alst.yaml [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+dataset_prepared_path: last_run_prepared
+val_set_size: 0.0
+output_dir: ./outputs/out
+
+sequence_len: 500_000
+min_sample_len: 200_000
+sample_packing: true
+
+tiled_mlp: true
+sequence_parallel_degree: 8
```
Comment (winglian) [inline examples/alst/llama3-8b-fsdp2-alst.yaml]: note to self that we'll need to change this to `context_parallel_size` when the other PR lands.

File: src/axolotl/monkeypatch/tiled_mlp/fsdp.py (hunk 1 above)
Comment (sfc-gh-truwase) [reply src/axolotl/monkeypatch/tiled_mlp/fsdp.py]: @winglian 

For z1 and z2: `hasattr(param, "param_idx_in_group")` 

For z3: `hasattr(param, "ds_id")`. 

It might be better to extend the z3 API to all deepspeed scenarios as the unified API. Please advise if that would be helpful.

Comment (sfc-gh-sbekman) [reply src/axolotl/monkeypatch/tiled_mlp/fsdp.py]: oh, so for z3 `param_idx_in_group` won't be there?

Comment (sfc-gh-sbekman) [reply src/axolotl/monkeypatch/tiled_mlp/fsdp.py]: IMHO, having some `.ds_blah` attribute that would indicate the model is wrapped in ZeRO would be most desirable for consistency.

like `.ds_zero` flag perhaps?

Comment (sfc-gh-truwase) [reply src/axolotl/monkeypatch/tiled_mlp/fsdp.py]: @sfc-gh-sbekman, I agreed with adding `.ds_zero` attribute. 

@winglian, will this work?

Comment (stas00) [reply src/axolotl/monkeypatch/tiled_mlp/fsdp.py]: which would be tricky as it'd require a new deepspeed release first, no?

Comment (sfc-gh-truwase) [reply src/axolotl/monkeypatch/tiled_mlp/fsdp.py]: Yes, new deepspeed release would be required. Until then, @winglian can continue with the separate detection logic for z1+z2 vs z3. That should work.

File: examples/alst/README.md [hunk 4]
```diff
@@ -0,0 +1,9 @@
+# Arctic Long Sequence Training (ALST)
+
+Artic Long Sequence Training (ALST) is a technique for training long context models using a variety of optimization
+techniques. It is a combination of:
+- TiledMLP: Leverage tiling over the sequence dimension on MLP layers to reduce memory usage
+- Tiled Loss: Using optimized loss functions like Liger-Kernel or Cut Cross Entropy to reduce memory usage
```
Comment (sfc-gh-sbekman) [inline examples/alst/README.md:6]:
```suggestion
- Fused Tiled Loss: Using optimized loss functions like Liger-Kernel or Cut Cross Entropy that don't manifest full logits to reduce memory usage
```

File: examples/alst/README.md [hunk 5]
```diff
@@ -0,0 +1,9 @@
+# Arctic Long Sequence Training (ALST)
+
+Artic Long Sequence Training (ALST) is a technique for training long context models using a variety of optimization
+techniques. It is a combination of:
+- TiledMLP: Leverage tiling over the sequence dimension on MLP layers to reduce memory usage
```
Comment (sfc-gh-sbekman) [inline examples/alst/README.md:5]:
```suggestion
- Ulysses SP: all-to-all-based sequence parallelism method
- TiledMLP: Leverage tiling over the sequence dimension on MLP layers to reduce memory usage
```

**Review by sfc-gh-sbekman — COMMENTED** [review]
a few small doc suggestions

File: examples/alst/README.md [hunk 6]
```diff
@@ -0,0 +1,10 @@
+# Arctic Long Sequence Training (ALST)
+
+Artic Long Sequence Training (ALST) is a technique for training long context models using a variety of optimization
```
Comment (sfc-gh-sbekman) [inline examples/alst/README.md:3]:
```suggestion
Arctic Long Sequence Training (ALST) is a technique for training long context models using a variety of optimization
```

File: src/axolotl/monkeypatch/tiled_mlp/fsdp.py (hunk 1 above)
Comment (winglian) [reply src/axolotl/monkeypatch/tiled_mlp/fsdp.py]: thanks @sfc-gh-truwase ! We'll use `param_idx_in_group` and `ds_id` checks for now
